diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1097 @@
+  enddo
+  fms_diag_buffer_init = allocated(buffobjs)
+end function
+
+!> creates a container type with a new (unallocated) buffer for the given dimensions
+function fms_diag_buffer_create_container(buff_dims) &
+result(rslt)
+  integer, intent(in)                            :: buff_dims !< dimensions 
+  type(fmsDiagBufferContainer_type), allocatable :: rslt
+
+  allocate(rslt)
+  select case (buff_dims)
+    case (0)
+      allocate(buffer0d :: rslt%diag_buffer_obj)
+    case (1)
+      allocate(buffer1d :: rslt%diag_buffer_obj)
+    case (2)
+      allocate(buffer2d :: rslt%diag_buffer_obj)
+    case (3)
+      allocate(buffer3d :: rslt%diag_buffer_obj)
+    case (4)
+      allocate(buffer4d :: rslt%diag_buffer_obj)
+    case (5)
+      allocate(buffer5d :: rslt%diag_buffer_obj)
+    case default
+      call mpp_error(FATAL, 'fms_diag_buffer_create_container: invalid number of dimensions given') 
+  end select
+end function 
+
+!!--------generic routines for any fmsDiagBuffer_class objects 
+
+!> setter for buffer_id for any buffer objects
+subroutine set_buffer_id(this, id)
+  class(fmsDiagBuffer_class), intent(inout) :: this !< buffer object to set id for
+  integer, intent(in)                       :: id !< positive integer id to set
+  if (.not.allocated(this%buffer_id) ) allocate(this%buffer_id)
+  this%buffer_id = id
+end subroutine
+
+!> Remaps 0-5d data buffer from the given object onto a 5d array and sets it in the type
+subroutine remap_buffer(this)
+  class(fmsDiagBuffer_class), intent(inout) :: this !< any dimension buffer object
+  integer                   :: buff_bounds(5)
+
+  if( DEBUG) print *, 'remapping buffer'
+  ! get num dimensions from type extension
+  select type (this)
+    type is (buffer0d)
+      if (.not. allocated(this%buffer)) call mpp_error(FATAL, "remap_buffer: buffer data not yet allocated")
+      if (DEBUG) print *, '0d buffer'
+      ! get buffer data type to allocate and remap
+      select type (buff => this%buffer)
+        type is(integer(i4_kind))
+          allocate (integer(kind=i4_kind) :: this%remap_buffer(1,1,1,1,1))
+          this%remap_buffer = buff
+        type is(integer(i8_kind))
+          allocate (integer(kind=i8_kind) :: this%remap_buffer(1,1,1,1,1))
+          this%remap_buffer = buff
+        type is(real(r4_kind))
+          allocate (real(kind=r4_kind) :: this%remap_buffer(1,1,1,1,1))
+          this%remap_buffer = buff
+        type is(real(r8_kind))
+          allocate (real(kind=r8_kind) :: this%remap_buffer(1,1,1,1,1))
+          this%remap_buffer = buff 
+      end select
+    type is (buffer1d)
+      if (.not. allocated(this%buffer)) call mpp_error(FATAL, "remap_buffer: buffer data not yet allocated")
+      if( DEBUG) print *, '1d buffer'
+      buff_bounds(1) = SIZE(this%buffer, 1)
+      select type (buff => this%buffer)
+        type is(integer(i4_kind))
+          allocate (integer(kind=i4_kind) :: this%remap_buffer(buff_bounds(1),1,1,1,1))
```

Comment:
I think remap_buffer should be a pointer.
Then all you will have to do is
```F90
this%remap_buffer(1:size(buff,1), 1:1, 1:1, 1:1, 1:1) => buff(1:size(buff,1))
```
you don't need the allocate and if you save the size of the buffer in the object you don't need the second select type because this works with class(*)

I don't think this will work for the buffer0d though. unless we do like this%buffer(1)

However either way this is not thread safe (@thomas-robinson?) since this%buffer will be shared by multiple threads, so the two threads will try to allocate/point remap_buffer. To make it thread safe(r), you will have to use the is_in, ie_in
https://github.com/NOAA-GFDL/FMS/blob/dmUpdate/diag_manager/diag_manager.F90#L1586 
```F90
this%remap_buffer(is:ie, 1:1, 1:1, 1:1, 1:1) => buff(is:ie)
```
That way each thread gets only the section of the variable that they need. 

or this will need to be inside an `!$OMP CRITICAL`

Reply:
This isn't something I really thought about yet.  We do need to consider how this will be affected by threading.  I wonder if there's a way to make each pointer private on a thread so they aren't sharing the pointer.

Reply:
@uramirez8707 and @rem1776 I *think* the problem is that we are storing the pointer in a module variable.  If we make the pointer-rank remapping local, it *should* be private, but then it's local, so we have to do whatever needs to be done in a local routine.

Reply:
@thomas-robinson  The pointer is in the buffer object type, not the module. I don't know what that would mean for the thread safety though. 

I was looking into the scalar remapping as well and I don't think its possible unless we do some C pointers. I was thinking maybe we could just do a "fake" scalar via a 1 element array.